src/pudl/output/eia860.py

```diff
@@ -464,3 +485,616 @@ def ownership_eia860(pudl_engine, start_date=None, end_date=None):
+    a single unit confidently.
+
+    Within each plant, for any generator that has no Unit ID, count the number
+    of generators of each prime mover type. Assign any collection of generators
+    having only a single CT or a single CA a unit ID, and flag it
+    "agg_unique_cc".
+
+    """
+    pm_cols = ["plant_id_eia", "generator_id", "prime_mover_code"]
+    # Get rid of the annual dependence, and only keep gens with missing PMs
+    df = gens_df[gens_df.unit_id_pudl.isna()][pm_cols].drop_duplicates()
+    # Count up the number of each kind of prime mover on a per-plant ID basis:
+    pm_counts = (
+        df.groupby(["plant_id_eia", "prime_mover_code"])
+        .size()
+        .unstack(fill_value=0)
+    )
+    # Plant IDs with at least one CT and one CA
+    # and either a single CA or a single CT.
+    # This makes the combined cycle unit grouping unambiguous.
+    cc_plant_ids = pm_counts[
+        (pm_counts["CA"] > 0)
+        & (pm_counts["CT"] > 0)
+        & ((pm_counts["CT"] == 1) | (pm_counts["CA"] == 1))
+    ].index
+    row_mask = (
+        gens_df.plant_id_eia.isin(cc_plant_ids)
+        & gens_df.unit_id_pudl.isnull()
+        & gens_df.prime_mover_code.isin(["CA", "CT"])
+    )
+
+    # Create Unit IDs for CA/CT generators that are being grouped together.
+    # Note that generator_id is not involved here because all of the
+    # selected generators within each plant are going to get the same Unit ID
+    unit_ids = (
+        gens_df.loc[row_mask, ["plant_id_eia", "unit_id_pudl"]]
+        .drop_duplicates()
+        .merge(
+            max_unit_id_by_plant(gens_df),
+            on="plant_id_eia",
+            how="left",
+            validate="many_to_one",
+        )
+        # Assign new unit_id_pudl values by incrementing beyond the previous max
+        .assign(
+            unit_id_pudl=lambda x: x.max_unit_id_pudl + 1,
+            bga_source=lambda x: "agg_unique_cc",
+        )
+        .drop("max_unit_id_pudl", axis="columns")
+    )
+
+    # Append the altered rows to the unaltered rows in a new dataframe to return
+    out_df = gens_df.loc[~row_mask].append(
+        gens_df.loc[row_mask]
+        .drop(["unit_id_pudl", "bga_source"], axis="columns")
+        .merge(
+            unit_ids,
+            on="plant_id_eia",
+            how="left",
+            validate="many_to_one",
+        )
+    )
+    return out_df
+
+
+def assign_complex_combined_cycle_unit_ids(gens_df):
+    """
+    Assign unit IDs to combined cycle plants that with multiple CT/CA PMs.
+
+    When there are multiple CA and CT units missing a Unit ID, we can't know
+    which ones should be associated with each other specifically, and so we
+    lump them all together into a single aggregated unit. In theory this should
+    yield reasonable heat rates that are the average of the multiple units that
+    might be getting lumped together.
+
+    """
+    pm_cols = ["plant_id_eia", "generator_id", "prime_mover_code"]
```

Comment:
bc moooost of the stuff (and the ideas between this function and `assign_unique_combined_cycle_unit_ids()` i miiiight think about squishing them together.

Reply:
IIRC, the only reason they're separate is so that we can differentiate between the two cases after the fact with different labels (since the unique ones ought to be higher confidence / granularity).  But yeah I think having a single function that does all of the Combined Cycle unit ID assignments would probably be fine and less duplicative.